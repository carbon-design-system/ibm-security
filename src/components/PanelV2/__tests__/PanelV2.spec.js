/**
 * @file Panel v2 tests.
 * @copyright IBM Security 2019
 */

import Add16 from '@carbon/icons-react/lib/add/16';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import { Button, PanelV2, PanelContent } from '../../..';

describe('PanelV2', () => {
  test('should have no Axe or DAP violations with custom footer via `renderFooter`', async () => {
    const main = document.createElement('main');
    render(
      <PanelV2
        title="test title"
        subtitle="test subtitle"
        closeButton={{
          label: 'test close',
        }}
        renderFooter={() => <Button>test footer button</Button>}
      >
        <PanelContent>test content</PanelContent>
      </PanelV2>,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations(
      'PanelV2 with renderFooter'
    );
  });

  test('should have no Axe or DAP violations with `title` or `subtitle` as a `node`', async () => {
    const main = document.createElement('main');
    render(
      <PanelV2
        title={<span>test title</span>}
        subtitle={<span>test subtitle</span>}
        closeButton={{
          label: 'test close',
        }}
      >
        <PanelContent>test content</PanelContent>
      </PanelV2>,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations(
      'PanelV2 with title or subtitle as node'
    );
  });

  test('should have no Axe or DAP violations when there is scrolling content', async () => {
    const main = document.createElement('main');
    render(
      <PanelV2
        // Note that title (or subtitle) should be provided here
        // to generate a valid `aria-labelledBy` for tabbable scrolling content:
        title="test title"
        closeButton={{
          label: 'test close',
        }}
        // Note that `hasScrollingContent={true}` so the `PanelContent` wrapper is tabbable:
        hasScrollingContent
        // `aria-label` provided because there is scrolling content:
        aria-label="test aria-label"
      >
        <PanelContent>
          test content text
          <Button>test content button</Button>
        </PanelContent>
      </PanelV2>,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations(
      'PanelV2 with hasScrollingContent'
    );
  });

  test('should have no Axe or DAP violations with deprecated `primaryButton` & `secondaryButton`', async () => {
    const main = document.createElement('main');
    render(
      <PanelV2
        title="test title"
        subtitle="test subtitle"
        closeButton={{
          label: 'test close',
        }}
        primaryButton={{
          icon: Add16,
          iconDescription: 'test icon description add',
          label: 'test primary button label',
        }}
        secondaryButton={{
          label: 'test secondary button label',
        }}
      >
        <PanelContent>test content</PanelContent>
      </PanelV2>,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations(
      'PanelV2 with deprecated props'
    );
  });

  test('should invoke close mock when close button is clicked', () => {
    const onCloseMock = jest.fn();
    const { getByLabelText } = render(
      <PanelV2
        closeButton={{
          label: 'test close',
          onClick: onCloseMock,
        }}
      />
    );

    userEvent.click(getByLabelText(/test close/i));
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  test('should not render panel when `isOpen` is set to `false`', () => {
    const { queryByText } = render(
      <PanelV2
        title="test title"
        isOpen={false}
        closeButton={{
          label: 'test close',
        }}
      />
    );

    expect(queryByText(/test title/i)).not.toBeInTheDocument();
  });

  test('should cycle panel elements in tab order', () => {
    const { getByLabelText, getByText } = render(
      <PanelV2
        closeButton={{
          label: 'test close',
        }}
        renderFooter={() => <Button>test footer button</Button>}
      >
        <PanelContent>
          test content text
          <Button>test content button</Button>
        </PanelContent>
      </PanelV2>
    );

    userEvent.tab();

    // The close button:
    expect(getByLabelText(/test close/i)).toHaveFocus();

    userEvent.tab();

    // The button inside the `PanelContent` wrapper:
    expect(getByText(/test content button/i)).toHaveFocus();

    userEvent.tab();

    // The footer button:
    expect(getByText(/test footer button/i)).toHaveFocus();

    userEvent.tab();

    // Loop complete.
    // The close button:
    expect(getByLabelText(/test close/i)).toHaveFocus();
  });

  test('should cycle panel elements in tab order when there is scrolling content', () => {
    const { getByLabelText, getByText } = render(
      <PanelV2
        closeButton={{
          label: 'test close',
        }}
        renderFooter={() => <Button>test footer button</Button>}
        // Note that `hasScrollingContent={true}` so the `PanelContent` wrapper is tabbable:
        hasScrollingContent
        // `aria-label` provided because there is scrolling content:
        aria-label="test aria-label"
      >
        <PanelContent>
          test content text
          <Button>test content button</Button>
        </PanelContent>
      </PanelV2>
    );

    userEvent.tab();

    // The close button:
    expect(getByLabelText(/test close/i)).toHaveFocus();

    userEvent.tab();

    // The tabbable `PanelContent` wrapper.
    // Need to use query selector that targets the accessible role
    // of the scrolling content due to divs wrapping inner text:
    expect(document.querySelector('section[role="region"]')).toHaveFocus();

    userEvent.tab();

    // The button inside the `PanelContent` wrapper:
    expect(getByText(/test content button/i)).toHaveFocus();

    userEvent.tab();

    // The footer button:
    expect(getByText(/test footer button/i)).toHaveFocus();

    userEvent.tab();

    // Loop complete.
    // The close button:
    expect(getByLabelText(/test close/i)).toHaveFocus();
  });
});
