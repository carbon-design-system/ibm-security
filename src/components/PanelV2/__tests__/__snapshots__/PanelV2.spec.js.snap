// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanelV2 renders 1`] = `
<PanelV2
  className={null}
  closeButton={
    Object {
      "onClick": [Function],
    }
  }
  focusTrap={true}
  isOpen={true}
  labels={Object {}}
  renderFooter={[Function]}
  stopPropagation={false}
  subtitle="Label"
  title="Label"
>
  <Transition
    className="security--panelv2"
    component={[Function]}
    timeout={160}
  >
    <CSSTransitionGroup
      className="security--transition"
      component={[Function]}
      transitionAppear={true}
      transitionAppearTimeout={160}
      transitionEnter={true}
      transitionEnterTimeout={160}
      transitionLeave={true}
      transitionLeaveTimeout={160}
      transitionName={
        Object {
          "appear": "security--panelv2__transition--appear",
          "appearActive": "security--panelv2__transition--appear--active",
          "enter": "security--panelv2__transition--enter",
          "enterActive": "security--panelv2__transition--enter--active",
          "leave": "security--panelv2__transition--leave",
          "leaveActive": "security--panelv2__transition--leave--active",
        }
      }
    >
      <TransitionGroup
        childFactory={[Function]}
        className="security--transition"
        component={[Function]}
        transitionAppear={true}
        transitionAppearTimeout={160}
        transitionEnter={true}
        transitionEnterTimeout={160}
        transitionLeave={true}
        transitionLeaveTimeout={160}
        transitionName={
          Object {
            "appear": "security--panelv2__transition--appear",
            "appearActive": "security--panelv2__transition--appear--active",
            "enter": "security--panelv2__transition--enter",
            "enterActive": "security--panelv2__transition--enter--active",
            "leave": "security--panelv2__transition--leave",
            "leaveActive": "security--panelv2__transition--leave--active",
          }
        }
      >
        <component
          className="security--transition"
        >
          <CSSTransitionGroupChild
            appear={true}
            appearTimeout={160}
            enter={true}
            enterTimeout={160}
            key=".$.0"
            leave={true}
            leaveTimeout={160}
            name={
              Object {
                "appear": "security--panelv2__transition--appear",
                "appearActive": "security--panelv2__transition--appear--active",
                "enter": "security--panelv2__transition--enter",
                "enterActive": "security--panelv2__transition--enter--active",
                "leave": "security--panelv2__transition--leave",
                "leaveActive": "security--panelv2__transition--leave--active",
              }
            }
          >
            <Portal
              focusTrap={true}
              hasOverlay={true}
              initialFocus={null}
              key=".0"
              rootNode={
                <body
                  class="security--portal__container"
                >
                  <section
                    class="security--panelv2 security--panelv2__transition--appear"
                  >
                    <header
                      class="security--panelv2__header"
                    >
                      <div
                        class="security--panelv2__header__container--title"
                      >
                        <h2
                          class="security--panelv2__header--title"
                        >
                          Label
                        </h2>
                        <div
                          class="security--panelv2__header--subtitle"
                        >
                          Label
                        </div>
                      </div>
                      <button
                        aria-label="Close"
                        class="security--button--icon security--panelv2__button--close"
                      >
                        <svg
                          aria-hidden="true"
                          class="security--icon"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          style="will-change: transform;"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                          />
                        </svg>
                      </button>
                    </header>
                    <section
                      class="security--panelv2__body security--panelv2__body--footer"
                      style="margin-top: 0px; margin-bottom: 0px;"
                    >
                      <div
                        class="security--panel__body--content"
                        role="tabpanel"
                        tabindex="0"
                      >
                        Label
                      </div>
                    </section>
                    <footer
                      class="security--panelv2__footer"
                    >
                      <button
                        class="security--button bx--btn bx--btn--secondary"
                        id="example-secondary-button"
                        tabindex="0"
                        type="button"
                      >
                        Close
                      </button>
                      <button
                        class="security--button bx--btn bx--btn--primary"
                        id="example-primary-button"
                        tabindex="0"
                        type="button"
                      >
                        Add
                      </button>
                    </footer>
                  </section>
                  <div
                    class="security--portal__overlay"
                    tabindex="-1"
                  />
                </body>
              }
              stopPropagation={false}
            >
              <Portal
                containerInfo={
                  <body
                    class="security--portal__container"
                  >
                    <section
                      class="security--panelv2 security--panelv2__transition--appear"
                    >
                      <header
                        class="security--panelv2__header"
                      >
                        <div
                          class="security--panelv2__header__container--title"
                        >
                          <h2
                            class="security--panelv2__header--title"
                          >
                            Label
                          </h2>
                          <div
                            class="security--panelv2__header--subtitle"
                          >
                            Label
                          </div>
                        </div>
                        <button
                          aria-label="Close"
                          class="security--button--icon security--panelv2__button--close"
                        >
                          <svg
                            aria-hidden="true"
                            class="security--icon"
                            focusable="false"
                            height="20"
                            preserveAspectRatio="xMidYMid meet"
                            style="will-change: transform;"
                            viewBox="0 0 32 32"
                            width="20"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                            />
                          </svg>
                        </button>
                      </header>
                      <section
                        class="security--panelv2__body security--panelv2__body--footer"
                        style="margin-top: 0px; margin-bottom: 0px;"
                      >
                        <div
                          class="security--panel__body--content"
                          role="tabpanel"
                          tabindex="0"
                        >
                          Label
                        </div>
                      </section>
                      <footer
                        class="security--panelv2__footer"
                      >
                        <button
                          class="security--button bx--btn bx--btn--secondary"
                          id="example-secondary-button"
                          tabindex="0"
                          type="button"
                        >
                          Close
                        </button>
                        <button
                          class="security--button bx--btn bx--btn--primary"
                          id="example-primary-button"
                          tabindex="0"
                          type="button"
                        >
                          Add
                        </button>
                      </footer>
                    </section>
                    <div
                      class="security--portal__overlay"
                      tabindex="-1"
                    />
                  </body>
                }
              >
                <FocusTrap
                  _createFocusTrap={[Function]}
                  active={true}
                  focusTrapOptions={
                    Object {
                      "initialFocus": null,
                    }
                  }
                  paused={false}
                >
                  <section
                    className="security--panelv2"
                  >
                    <header
                      className="security--panelv2__header"
                    >
                      <div
                        className="security--panelv2__header__container--title"
                      >
                        <h2
                          className="security--panelv2__header--title"
                        >
                          Label
                        </h2>
                        <div
                          className="security--panelv2__header--subtitle"
                        >
                          Label
                        </div>
                      </div>
                      <IconButton
                        className="security--panelv2__button--close"
                        iconClassName=""
                        iconSize={20}
                        label="Close"
                        onClick={[Function]}
                        path={null}
                        renderIcon={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "render": [Function],
                          }
                        }
                        size={null}
                        state={false}
                        tooltip={false}
                        tooltipDirection="bottom"
                      >
                        <button
                          aria-label="Close"
                          className="security--button--icon security--panelv2__button--close"
                          onClick={[Function]}
                        >
                          <Icon
                            className=""
                            height={null}
                            path={null}
                            renderIcon={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "render": [Function],
                              }
                            }
                            size={20}
                            viewBox="0 0 32 32"
                            width={null}
                          >
                            <ForwardRef(Close20)
                              aria-hidden="true"
                              className="security--icon"
                              focusable={false}
                              preserveAspectRatio="xMidYMid meet"
                              style={
                                Object {
                                  "willChange": "transform",
                                }
                              }
                            >
                              <Icon
                                aria-hidden="true"
                                className="security--icon"
                                focusable={false}
                                height={20}
                                preserveAspectRatio="xMidYMid meet"
                                style={
                                  Object {
                                    "willChange": "transform",
                                  }
                                }
                                viewBox="0 0 32 32"
                                width={20}
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <svg
                                  aria-hidden={true}
                                  className="security--icon"
                                  focusable={false}
                                  height={20}
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "willChange": "transform",
                                    }
                                  }
                                  viewBox="0 0 32 32"
                                  width={20}
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                                  />
                                </svg>
                              </Icon>
                            </ForwardRef(Close20)>
                          </Icon>
                        </button>
                      </IconButton>
                    </header>
                    <section
                      className="security--panelv2__body security--panelv2__body--footer"
                      style={
                        Object {
                          "marginBottom": "0px",
                          "marginTop": "0px",
                        }
                      }
                    >
                      <PanelContent>
                        <div
                          className="security--panel__body--content"
                          role="tabpanel"
                          tabIndex="0"
                        >
                          Label
                        </div>
                      </PanelContent>
                    </section>
                    <footer
                      className="security--panelv2__footer"
                    >
                      <Button
                        disabled={false}
                        id="example-secondary-button"
                        kind="secondary"
                        largeText={null}
                        loading={false}
                        onClick={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        <ForwardRef(Button)
                          className="security--button"
                          disabled={false}
                          id="example-secondary-button"
                          kind="secondary"
                          onClick={[Function]}
                          tabIndex={0}
                          type="button"
                        >
                          <button
                            className="security--button bx--btn bx--btn--secondary"
                            disabled={false}
                            id="example-secondary-button"
                            onClick={[Function]}
                            tabIndex={0}
                            type="button"
                          >
                            Close
                          </button>
                        </ForwardRef(Button)>
                      </Button>
                      <Button
                        disabled={false}
                        id="example-primary-button"
                        kind="primary"
                        largeText={null}
                        loading={false}
                        onClick={[Function]}
                        tabIndex={0}
                        type="button"
                      >
                        <ForwardRef(Button)
                          className="security--button"
                          disabled={false}
                          id="example-primary-button"
                          kind="primary"
                          onClick={[Function]}
                          tabIndex={0}
                          type="button"
                        >
                          <button
                            className="security--button bx--btn bx--btn--primary"
                            disabled={false}
                            id="example-primary-button"
                            onClick={[Function]}
                            tabIndex={0}
                            type="button"
                          >
                            Add
                          </button>
                        </ForwardRef(Button)>
                      </Button>
                    </footer>
                  </section>
                </FocusTrap>
              </Portal>
            </Portal>
          </CSSTransitionGroupChild>
        </component>
      </TransitionGroup>
    </CSSTransitionGroup>
  </Transition>
</PanelV2>
`;
